File: DataLoggerHTML2CSVParser.py
#!/usr/bin/env python3

import pathlib as PL
import pandas as PD
from bs4 import BeautifulSoup as BS # I adore this terminology.

#' read a vusitu html file and extract meta info and data
def ReadVuSituHTML(file_path: PL.Path, meta_selection: list = None) -> PD.DataFrame:

    # read file content
    with open(file_path) as fi:
        file_content = BS(fi, "html.parser")
    # print(file_content)
    # data_table = PD.read_html(file_path) # has meta info in table

    ### meta data
    if meta_selection is None:
        meta_selection = ["SerialNumber", "StartTime", "Duration"]

    meta_infos = []
    for mesel in meta_selection:
        info = file_content.find("td", attrs = {"isi-property": mesel}).text
        meta_infos.append(info.split(" = "))

    meta_infos = {k: v for k, v in meta_infos}


    ### data table
    extract_row = lambda row: [td.get_text() for td in row.find_all("td")]

    # header
    header = extract_row(file_content.find("tr", {"class" : "dataHeader"}))

    # data types, used further down
    data_types = file_content.find("tr", {"class" : "dataHeader"}).find_all("td")
    data_types = {
        col: dt for col, dt
        in zip(header, [dt["isi-data-column-header"] for dt in data_types])
    }

    # rowwise data extraction
    data_rows = [
        extract_row(datarow)
        for datarow
        in file_content.find_all("tr", {"class" : "data"})
        ]

    # DataFrame conversion
    data_dict = {
        hd: dat
        for hd, dat in
        list(zip(header, zip(*data_rows)))
        }

    data_table = PD.DataFrame.from_dict(data_dict)

    # data type conversion
    timestamps = [col for col in data_table.columns if data_types[col] == "DateTime"]
    for col in timestamps:
        data_table.loc[:, col ] = PD.to_datetime(data_table.loc[:, col].values)

    # others are float
    data_table.astype({
        col: "float" for col in data_table.columns
        if data_types[col] == "Parameter"
    })
    # print(data_table.dtypes)

    ### done
    return meta_infos, data_table



if __name__ == "__main__":

    # testing
    data_path = PL.Path("test_data")
    test_file = data_path / "VuSitu_Loggen_2025-08-06_12-00-00_Default_Site_BABP042X.html"
    test_output = ReadVuSituHTML(
        test_file,
        meta_selection = ["SerialNumber", "StartTime", "Duration", "FirmwareVersion"]
    )
    print(test_output[0])
    print(test_output[1].head(5))

File: DataLoggerHTML2CSVParser.R
#!/usr/bin/Rscript

# # Load the packages
# library("rvest")
# library("xml2")
# library("dplyr")
# library("janitor") # to cleanup the data

# this is loading a test file
test_file <- here::here(
  "test_data",
  "VuSitu_Loggen_2025-08-06_12-00-00_Default_Site_BABP042X.html"
  )


#' read a vusitu html file and extract meta info and data
#'
#' @param file_path a path to the file which shall be read
#' @param meta_selection character array of meta data tags to read
#' @return a list with the elements "meta" (list) and "data" (tibble)
#'
read_vusitu_html <- function(file_path, meta_selection = NULL) {

  # does file exist?
  if (isFALSE(file.exists(file_path))) stop("File not found.")

  # libraries
  stopifnot(
    "rvest" = require("rvest"),
    "xml2" = require("xml2"),
    "dplyr" = require("dplyr"),
    "janitor" = require("janitor")
  )

  # read in the raw html
  # https://www.geeksforgeeks.org/web-scraping/scrape-an-html-table-using-rvest-in-r
  logger_readout_raw <- rvest::read_html(file_path)

  # get meta node content
  meta_nodes <- logger_readout_raw %>%
    rvest::html_elements(".sectionMember") %>%
    rvest::html_nodes("td")

  if (is.null(meta_selection)) {
    meta_selection <- c("SerialNumber", "StartTime", "Duration")
  }

  select_node <- function(isi_properties) meta_nodes %>%
    rvest::html_attr("isi-property") %in% isi_properties
  meta_info <- meta_nodes[select_node(meta_selection)] %>%
    rvest::html_text()

  # format into named list
  meta_info <- sapply(meta_info, FUN = function(info) strsplit(info, " = "))
  names(meta_info) <- sapply(meta_info, `[[`, 1)
  meta_info <- sapply(meta_info, `[[`, 2)

  # # in principle, this can get only data nodes,
  # but I was unsuccesful to use them further
  # logger_readout_raw %>% html_nodes(".data")

  # find the raw table
  table_raw <- logger_readout_raw %>% rvest::html_elements("table")


  # find rows of the correct classes ("data" and "dataHeader")
  data_rows <- unlist(
      lapply(
        table_raw %>% rvest::html_elements("tr"),
        FUN = function(tr) tr %>% rvest::html_attr("class")
      )
    ) %>% startsWith("data")
  data_rows[is.na(data_rows)] <- FALSE # NANs are not data rows

  # this assumes that only one table is present
  table_data <- table_raw %>%
    rvest::html_table() %>%
    .[[1]] %>%
    .[data_rows,] %>%
    dplyr::filter(dplyr::if_all(dplyr::everything(), ~ !is.na(.x))) %>%
    janitor::row_to_names(row = 1) # first row is dataHeader

  # data type cleanup
  table_data <- table_data %>%
    dplyr::mutate_at(
      dplyr::vars(dplyr::starts_with("Datum")),
      function(ts) as.POSIXct(ts, format = "%Y-%m-%d %H:%M:%OS")
      ) %>%
    mutate_at(
      dplyr::vars(dplyr::starts_with(c("Temperatuur", "Diepte", "Druk"))),
      ~as.numeric(.) # ~readr::parse_number(.)
      )

  return(list(
    "meta" = meta_info,
    "data" = table_data
  ))
} # /read_vusitu_html

# options(digits = 12, digits.secs = 4)
test_output <- read_vusitu_html(
  file_path = test_file,
  meta_selection = c("SerialNumber", "StartTime", "Duration", "FirmwareVersion")
)
test_output["meta"] %>% knitr::kable(col.names = NULL, format = "pandoc")
test_output["data"] %>% knitr::kable(digits = Inf)
# test_output %>% pull("Datum Tijd") %>% format("%Y%m%d %H:%M:%OS6")
